p2p/src/codec.rs

```diff
@@ -0,0 +1,244 @@
+		Ok(())
+	}
+
+	fn read_inner(&mut self) -> Result<Message, Error> {
+		self.bytes_read = 0;
+		loop {
+			let next_len = self.next_len();
+			let pre_len = self.buffer.len();
+			// Buffer could already be partially filled, calculate additional bytes we need
+			let to_read = next_len.saturating_sub(pre_len);
+			if to_read > 0 {
+				self.buffer.reserve(to_read);
+				for _ in 0..to_read {
+					self.buffer.put_u8(0);
+				}
+				self.set_stream_timeout()?;
+				if let Err(e) = self.stream.read_exact(&mut self.buffer[pre_len..]) {
+					// Undo reserved bytes on a failed read
+					self.buffer.truncate(pre_len);
+					return Err(e.into());
+				}
+				self.bytes_read += to_read;
+			}
+			match &mut self.state {
+				None => {
+					// Parse header and keep reading
+					let mut raw = self.buffer.split_to(next_len).freeze();
+					let mut reader = BufReader::new(&mut raw, self.version);
+					let header = MsgHeaderWrapper::read(&mut reader)?;
+					self.state = Header(header);
+				}
+				Header(Known(header)) => {
+					let mut raw = self.buffer.split_to(next_len).freeze();
+					if header.msg_type == Type::Headers {
+						// Special consideration for a list of headers, as we want to verify and process
+						// them as they come in instead of only after the full list has been received
+						let mut reader = BufReader::new(&mut raw, self.version);
+						let items_left = reader.read_u16()? as usize;
+						self.state = BlockHeaders {
+							bytes_left: header.msg_len as usize - 2,
+							items_left,
+							headers: Vec::with_capacity(min(HEADER_BATCH_SIZE, items_left)),
+						};
+					} else {
+						// Return full message
+						let msg = decode_message(header, &mut raw, self.version);
+						self.state = None;
+						return msg;
+					}
+				}
+				Header(Unknown(_, msg_type)) => {
+					// Discard body and return
+					let msg_type = *msg_type;
+					self.buffer.advance(next_len);
+					self.state = None;
+					return Ok(Message::Unknown(msg_type));
+				}
+				BlockHeaders {
+					bytes_left,
+					items_left,
+					headers,
+				} => {
+					if *bytes_left == 0 {
+						// Incorrect item count
+						self.state = None;
+						return Err(Error::BadMessage);
+					}
+
+					let mut reader = BufReader::new(&mut self.buffer, self.version);
+					let header: UntrustedBlockHeader = reader.body()?;
+					let bytes_read = reader.bytes_read() as usize;
+					headers.push(header.into());
+					*bytes_left = bytes_left.saturating_sub(bytes_read);
+					*items_left -= 1;
+
+					if headers.len() == HEADER_BATCH_SIZE || *items_left == 0 {
+						let mut h = Vec::with_capacity(min(HEADER_BATCH_SIZE, *items_left));
+						mem::swap(headers, &mut h);
+						if *items_left == 0 {
+							let bytes_left = *bytes_left;
+							self.state = None;
+							if bytes_left > 0 {
+								return Err(Error::BadMessage);
+							}
+						}
+						return Ok(Message::Headers(h));
+					}
+				}
+				Attachment(left, meta, now) => {
```

Comment:
And in contrast with headers, reading an attachment returns a _single_ `Message`, basically a running total of the download? Each time with potentially more of the attachment read? 
Subsequent blocking calls to `read_inner()` may result in more of the attachment being read.

Reply:
An attachment will generate a lot of `Message::Attachment`s, one for every 48000 bytes that is read. This is the same size iirc as the buffer we were filling before.